functionary/prompt_template/llama3_prompt_template.py

```diff
@@ -0,0 +1,301 @@
+import json
+import random
+import string
+from typing import Any, Dict, List, Literal, Optional, Tuple, Union
+
+from functionary.prompt_template import prompt_utils
+from functionary.prompt_template.base_template import PromptTemplate
+
+
+def convert_to_llama3_messages(messages: List[Dict]) -> List[Dict]:
+    result = []
+    index = 0
+    while index < len(messages):
+        if messages[index]["role"] in ["user", "system"]:
+            result.append(messages[index])
+            index += 1
+        else:
+            if messages[index]["role"] == "assistant":
+                tool_calls = messages[index].get("tool_calls", [])
+                if len(tool_calls) == 0:
+                    result.append(messages[index])
+                else:
+                    messages
+
+
+class Llama3Template(PromptTemplate):
+    function_separator = "<|reserved_special_token_249|>"
+    version = "v2.llama3"
+    fn_param_sep_token = "\n"
+
+    def get_additional_tokens(self) -> List[str]:
+        return []
+
+    def get_assistant_prefixes(self) -> List[str]:
+        # return [f"{self.from_token}assistant\n{self.recipient_token}"]
+        return ["<|start_header_id|>assistant<|end_header_id|>\n\n"]
+
+    def get_stop_tokens_for_generation(self) -> List[str]:
+        return ["<|eot_id|>", "<|end_of_text|>"]
```

Comment:
Seems like meta is not using `<|end_of_text|>` already

Reply:
Yeah, I added this just in case, adding is better than removing